include/psa/crypto_values.h

```diff
@@ -2117,7 +2297,8 @@ static inline int mbedtls_svc_key_id_is_null( mbedtls_svc_key_id_t key )
 /** A salt for key derivation.
  *
  * This should be a direct input.
- * It can also be a key of type #PSA_KEY_TYPE_RAW_DATA.
+ * It can also be a key of type #PSA_KEY_TYPE_RAW_DATA or
+ * #PSA_KEY_TYPE_PEPPER.
```

Comment:
Argon2 officially takes both a salt and a pepper. For other algorithms, when a pepper is used, it's a part of the salt, with no particular standard to put it first or last.

Reply:
> Our first priority is PBKDF2. If pepper needs tuning, we can deal with it later.

So - should we remove `PSA_KEY_TYPE_PEPPER` until we need it for another algorithm?

Reply:
Now I just remembered the discussion that we already had about salt and pepper: typical usage for algorithms like PBKDF2 is to concatenate the per-application secret pepper and per-id non-secret salt and provide them as the "salt" to PBKDF2.

Hence using the INPUT_SALT with a new key type, mostly to help the developer pay attention to what the data in the "key" is. The proposed API design just reflects the evolved use (salt+pepper) of existing password hashing functions (which only knew about salt).

The API might still look cleaner if we provided as `INPUT_PEPPER` step (for use with `PSA_KEY_TYPE_PEPPER` keys), and algorithms such as PBKDF2 documented that any `INPUT_PEPPER` inputs are concatenated in sequence with any `INPUT_SALT` inputs?

Reply:
No, we should keep pepper: we know it's useful, including for PBKDF2. And yes, for most algorithms including PBKDF2, pepper and salt are the same input. Since a pepper is secret, it should have its own type.

Reply:
Ok, so it seems we agree about keeping `KEY_TYPE_PEPPER`.

I think there's a separate question of whether we should use `INPUT_PEPPER` for PBKDF2. (I think it's quite clear that when we add Argon2, it will have `INPUT_PEPPER`, so below I'm just focusing on the question of introducing it for things like PBKDF2.) I think we need to have a special case somewhere, and introducing `INPUT_PEPPER` or not determines where this special case lives.

1. If we introduce `INPUT_PEPPER`, the nice thing is that `INPUT_SALT` keeps using `KEY_TYPE_RAW_DATA` like all the other non-secret inputs. The special case is that (for all algs except Argon2), `INPUT_PEPPER` is not a separate input but is concatenated with other `INPUT_SALT` inputs; and we still have to document that `INPUT_SALT` can be repeated (in case some scheme uses a salt structure like `per_entry_salt || pepper || context`). Also, the list of input steps accepted by PBKDF2 will look different to the descriptions of its input in the standard.
2. If we don't introduce `INPUT_PEPPER`, the nice thing is that the description of the input steps matches the standard. The special case is that `INPUT_SALT` must be documented as accepting both secret inputs (with `KEY_TYPE_PEPPER`) and non secret inputs (with `KEY_TYPE_RAW_DATA` or more likely `input_bytes()`). (Plus that `INPUT_SALT` can be repeated, but that's actually common with the other case.)

I think option 2, which is the current state of the PR, better represents the standard (input steps match it) + usage that evolved on top of it (key type tells the _nature_ of the data passed at this step). Though I'm open to hearing arguments for option 1 if you're not convinced!

Reply:
We discussed in https://github.com/ARMmbed/psa-crypto-api/pull/489 and came to the conclusion that it was best to stick to `INPUT_SALT` for algorithms that don't have a dedicated pepper input, because the pepper would be concatenated into the salt. I don't see any new arguments in this discussion, so let's stick to what we have.